File: examples/typescript-api/src/domain/user/repository.ts
import { User } from './entity';

/**
 * User Repository Interface
 *
 * 도메인 레이어에서 정의하는 저장소 인터페이스입니다.
 * 실제 구현은 Infrastructure 레이어에서 합니다.
 */
export interface IUserRepository {
  findById(id: string): Promise<User | null>;
  findByEmail(email: string): Promise<User | null>;
  save(user: User): Promise<void>;
}

File: examples/python-api/app/domain/user/repository.py
"""User Repository Interface."""

from typing import Protocol

from app.domain.user.entity import User


class UserRepository(Protocol):
    """Domain Layer가 정의하는 Repository 인터페이스.
    Infrastructure Layer에서 구현합니다."""

    async def find_by_id(self, user_id: str) -> User | None: ...
    async def find_by_email(self, email: str) -> User | None: ...
    async def save(self, user: User) -> None: ...
